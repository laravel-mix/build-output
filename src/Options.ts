/**
 * Control what and how the table is written to the console
 **/
export interface Options {
  /**
   * Clear the console when rendering the table
   *
   * Defaults to `false`
   */
  clearConsole?: boolean

  /**
   * Show related assets (when available) in the table
   *
   * Defaults to `false`
   */
  showRelated?: boolean

  /**
   * The header to print (if any)
   *
   * Defaults to `undefined`
   */
  header?: string

  /**
   * Attempt to fit the table within this width if possible
   *
   * Defaults to the width of the current terminal window
   */
  width?: number

  /**
   * The maximum number of columns a table can occupy
   *
   * Defaults to `Infinity`
   */
  maxWidth?: number

  /**
   * Whether or not to use colors when printing
   *
   * Defaults to `undefined` which auto-detects colors
   */
  colors?: true | false
}
