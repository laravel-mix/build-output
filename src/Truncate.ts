/**
 * Customize how the string is truncated
 *
 * @internal
 **/
export interface TruncateOptions {
  length: number
  location?: "start" | "middle" | "end"
  omission?: string
}

/**
 * A somewhat unicode-aware truncation algorithm that can truncate
 * a string to a given length at the start, middle, or end.
 *
 * @internal
 **/
export function truncate(str: string, options: TruncateOptions) {
  options.location = options.location ?? "end"
  options.omission = options.omission ?? "â€¦"

  const strChars = [...str]
  const omissionChars = [...options.omission]
  const maxLength = options.length - omissionChars.length

  if (strChars.length <= options.length) {
    return str
  }

  if (maxLength < 0) {
    return options.omission
  }

  // Based on the location we need to figure out where to slice the original string
  const result: string[] = []

  if (options.location === "start") {
    result.push(...omissionChars)
    result.push(...strChars.slice(strChars.length - maxLength))
  }

  if (options.location === "middle") {
    result.push(...strChars.slice(0, maxLength / 2))
    result.push(...omissionChars)
    result.push(...strChars.slice(strChars.length - maxLength / 2))
  }

  if (options.location === "end") {
    result.push(...strChars.slice(0, maxLength))
    result.push(...omissionChars)
  }

  return result.join("")
}
