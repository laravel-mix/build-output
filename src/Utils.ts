/**
 * Turn a number of bytes into a human readable size
 *
 * For instance:
 * - 1000 -> 1KB
 * - 3745100 -> 3.745MB
 *
 * @param {number} size the size in bytes
 * @returns {string} the formatted size
 * @internal
 */
export function formatSize(size: number): string {
  if (Number.isNaN(size) === true) {
    return "unknown"
  }

  const abbreviations = ["bytes", "KB", "MB", "GB", "TB"]

  const index = Math.floor(Math.log(size) / Math.log(10 ** 3))
  const amount = size / 1000 ** index

  const amountStr = index === 0 ? `${amount}` : `${amount.toPrecision(3)}`

  return `${amountStr} ${abbreviations[index]}`
}
