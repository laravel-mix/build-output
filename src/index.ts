export * from "./Options"
export * from "./BuildOutputPlugin"
export { BuildOutputPlugin as default } from "./BuildOutputPlugin"
