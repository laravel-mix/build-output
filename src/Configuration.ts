import { Colors } from "picocolors/types"
import { ConsoleOutput } from "./Console"
import { Options } from "./Options"

/**
 * This configuration object is passed around to any components that need access to the configuration information
 * If that component itself is stored in the config object it must be optional
 *
 * @internal
 **/
export class Configuration {
  constructor(public options: Options, public colors: Colors, public console?: ConsoleOutput) {}
}
