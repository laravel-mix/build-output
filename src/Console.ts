import readline from "readline"
import { WriteStream } from "tty"
import { Configuration } from "./Configuration"

/**
 * Represents the capabilities to read information about and write data to a console/terminal/etc…
 *
 * @public
 **/
export interface ConsoleOutput {
  get width(): number
  get height(): number

  clear(): void
  heading(text: string): void
  success(text: string): void
  write(str: string): void
}

/**
 * The default internal console implementation that writes to a given stream like `process.stdout`
 *
 * @internal
 **/
export class Console {
  constructor(private configuration: Configuration, private stream: WriteStream) {}

  public get width() {
    return this.stream.columns
  }

  public get height() {
    return this.stream.rows
  }

  public clear() {
    this.write("\n".repeat(this.height))
    readline.cursorTo(this.stream, 0, 0)
    readline.clearScreenDown(this.stream)
  }

  /**
   * Print a block section heading.
   */
  public heading(text: string) {
    this.write("\n")
    this.write(this.colors.bgBlue(this.colors.bgBlue(this.colors.bold(this.section(text)))))
    this.write("\n")
  }

  /**
   * Print a success message
   */
  public success(text: string) {
    this.write(this.colors.green(this.colors.bold(`✔ ${text}`)))
    this.write("\n")
  }

  /**
   * Write directly to the console
   */
  public write(str: string) {
    this.stream.write(str)
  }

  private get colors() {
    return this.configuration.colors
  }

  /**
   * Create a block section.
   */
  private section(text: string) {
    const padLength = 3
    const padding = " ".repeat(padLength)

    text = `${padding}${text}${padding}`

    const line = " ".repeat(text.length)

    return `${line}\n${text}\n${line}`
  }
}
