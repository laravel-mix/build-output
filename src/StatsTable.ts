import stripAnsi from "strip-ansi"
import { getBorderCharacters, table, TableUserConfig } from "table"
import { StatsCompilation } from "webpack"
import { formatSize } from "./Utils"
import { Configuration } from "./Configuration"
import { truncate } from "./Truncate"

/**
 * Generate a table representing assets from webpack stats
 *
 * @internal
 **/
export class StatsTable {
  constructor(private configuration: Configuration) {}

  /**
   * Return a table representing the compilation stats as a string
   */
  public render(data: StatsCompilation): string {
    const rows: string[][] = []

    rows.push([this.colors.bold("File"), this.colors.bold("Size")])

    const assets = this.sortAssets(data)

    for (const asset of assets) {
      rows.push([this.colors.green(asset.name), formatSize(asset.size)])
    }

    return this.createTable(rows)
  }

  private sortAssets(data: StatsCompilation) {
    let assets = data.assets!.flatMap((asset) => [asset, ...(Array.isArray(asset.related) ? asset.related : [])])

    assets = assets.sort((a, b) => {
      return a.name.localeCompare(b.name) || a.size - b.size
    })

    return assets
  }

  private get colors() {
    return this.configuration.colors
  }

  private get options() {
    return this.configuration.options
  }

  private get console() {
    return this.configuration.console!
  }

  private createTable(rows: string[][]) {
    const tableConfig: TableUserConfig = {
      border: getBorderCharacters("norc"),

      columns: [
        {width: 35, alignment: "right"},
        {alignment: "left"},
      ],

      drawHorizontalLine(index) {
        // Top of table == index 0
        // Bottom table header == index 1
        // Bottom table == index rows.length

        return index <= 1 || index === rows.length
      },
    }

    const toTable = () => table(rows, tableConfig).trim()
    const calculateWidth = () => stripAnsi(toTable().split('\n')[0]).length

    const targetWidth = this.options.width ?? this.console.width
    const maxWidth = this.options.maxWidth ?? Infinity

    let currentWidth = calculateWidth()
    let fileColIncrease = Math.min(targetWidth - currentWidth, maxWidth - currentWidth)

    // If the table doesn't fit within the desired width then we increase the width of the file path column
    if (fileColIncrease > 0) {
      // @ts-ignore
      tableConfig.columns![0]!.width! += fileColIncrease
    }

    // start-truncate the path names
    for (const row of rows) {
      row[0] = truncate(row[0], { length: tableConfig.columns![0]!.width!, location: "start" })
    }

    return toTable()
  }
}
