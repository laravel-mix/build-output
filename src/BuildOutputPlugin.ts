import { createColors } from "nanocolors"
import { Compiler, Stats } from "webpack"
import { ConsoleOutput, Console } from "./Console"
import { Configuration } from "./Configuration"
import { Options } from "./Options"
import { StatsTable } from "./StatsTable"

/**
 * Add fancy a table to your webpack build output
 *
 * @public
 **/
export class BuildOutputPlugin {
  private configuration: Configuration

  /** @internal */
  public constructor(config: Configuration)

  public constructor(options: Options, console?: ConsoleOutput)

  public constructor(
    options?: Options | Configuration,
    console?: ConsoleOutput,
  ) {
    if (! (options instanceof Configuration)) {
      options = this.resolveConfig(options ?? {}, console)
    }

    if (! options.console) {
      throw new Error("Custom configurations must provide a console writer to use")
    }

    this.configuration = options
  }

  public apply(compiler: Compiler) {
    compiler.hooks.done.tap("BuildOutputPlugin", (stats) => this.render(stats))
  }

  // Public only for testing purposes
  /** @private */
  public render(stats: Stats) {
    if (stats.hasErrors()) {
      return false
    }

    const data = stats.toJson({
      assets: true,
      builtAt: true,
      hash: true,
      performance: true,
      relatedAssets: this.options.showRelated,
    })

    if (this.options.clearConsole) {
      this.console.clear()
    }

    // In case the header is defined as a get property this will ensure to only access it once
    const header = this.options.header

    if (header) {
      this.console.heading(header)
    }

    this.console.success(`Compiled Successfully in ${data.time}ms`)

    if (data.assets?.length) {
      this.console.write(new StatsTable(this.configuration).render(data))
    }
  }

private resolveConfig(options: Options, console?: ConsoleOutput) {
    const config = new Configuration(
      options,
      createColors(options.colors)
    )

    config.console = console ?? new Console(config, process.stdout)

    return config
  }

  private get options() {
    return this.configuration.options
  }

  private get console() {
    return this.configuration.console!
  }
}
