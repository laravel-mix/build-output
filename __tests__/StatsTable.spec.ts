import { StatsTable } from "../src/StatsTable"
import { createFakeConfig } from "./utils/Fake"
import { asset } from "./utils/Webpack"

test("The stats table renders as expected", () => {
  const table = new StatsTable(createFakeConfig())

  const output = table.render({
    assets: [
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/app.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/bar.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/extracted.css",
        size: 1500,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/app.css",
        size: 1000,
      }),
    ],
  })

  expect(`\n${output}\n`).toMatchSnapshot()
})

test("The stats table can be given a maximum length", () => {
  const table = new StatsTable(createFakeConfig({ maxWidth: 50 }))

  const output = table.render({
    assets: [
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/app.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/bar.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/extracted.css",
        size: 1500,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/app.css",
        size: 1000,
      }),
    ],
  })

  expect(`\n${output}\n`).toMatchSnapshot()
})
