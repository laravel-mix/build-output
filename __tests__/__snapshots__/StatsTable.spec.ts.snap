// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stats table can be given a maximum length 1`] = `
"
┌──────────────────────────────────────┬─────────┐
│                                 File │ Size    │
├──────────────────────────────────────┼─────────┤
│ …foo/foo/foo/foo/foo/foo/css/app.css │ 1.00 KB │
│ …o/foo/foo/foo/foo/css/extracted.css │ 1.50 KB │
│ …o/foo/foo/foo/foo/foo/foo/js/app.js │ 1.20 KB │
│ …o/foo/foo/foo/foo/foo/foo/js/bar.js │ 1.20 KB │
└──────────────────────────────────────┴─────────┘
"
`;

exports[`The stats table renders as expected 1`] = `
"
┌────────────────────────────────────────────────────────────────────┬─────────┐
│                                                               File │ Size    │
├────────────────────────────────────────────────────────────────────┼─────────┤
│                            foo/foo/foo/foo/foo/foo/foo/css/app.css │ 1.00 KB │
│                      foo/foo/foo/foo/foo/foo/foo/css/extracted.css │ 1.50 KB │
│                              foo/foo/foo/foo/foo/foo/foo/js/app.js │ 1.20 KB │
│                              foo/foo/foo/foo/foo/foo/foo/js/bar.js │ 1.20 KB │
└────────────────────────────────────────────────────────────────────┴─────────┘
"
`;
